lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,313 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2021 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of shared script state using
+-- scheme specified in CIP-XXX Multi-signature Wallets.
+
+module Cardano.Wallet.Primitive.AddressDiscovery.Shared
+    (
+    -- ** State
+      SharedState (..)
+    , unsafeSharedState
+    , newSharedState
+    , extendSharedState
+    , addCosignerAccXPub
+    , purposeCIPXXX
+    , isShared
+    , keyHashFromAccXPubIx
+    , constructAddressFromIx
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( Cosigner (..)
+    , KeyHash (..)
+    , Script (..)
+    , ScriptTemplate (..)
+    , ValidationLevel (..)
+    , foldScript
+    , toScriptHash
+    , validateScriptTemplate
+    )
+import Cardano.Address.Style.Shelley
+    ( Credential (..)
+    , delegationAddress
+    , deriveMultisigPublicKey
+    , hashKey
+    , liftXPub
+    , mkNetworkDiscriminant
+    )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationPrefix (..)
+    , DerivationType (..)
+    , Index (..)
+    , NetworkDiscriminant (..)
+    , SoftDerivation
+    , WalletKey (..)
+    , deriveVerificationKey
+    , hashVerificationKey
+    )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( coinTypeAda )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( AddressPoolGap )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..), AddressState (..) )
+import Control.DeepSeq
+    ( NFData )
+import Data.Either
+    ( isRight )
+import Data.Map.Strict
+    ( Map )
+import Data.Maybe
+    ( fromMaybe )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Address as CA
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+                                 State
+-------------------------------------------------------------------------------}
+
+-- | A state to keep track of script templates, account public keys of other cosigners,
+-- | verification keys used and unused fitting in the address pool gap,
+-- | and script addresses unused/used with the corresponding indices.
+data SharedState (n :: NetworkDiscriminant) k = SharedState
+    { shareStateAccountKey :: k 'AccountK XPub
+        -- ^ Reward account public key associated with this wallet
+    , shareStateDerivationPrefix :: !DerivationPrefix
+        -- ^ Derivation path prefix from a root key up to the account key
+    , shareStateGap :: !AddressPoolGap
+        -- ^ Number of first keys that are used to produce candidate addresses
```

Comment:
The comment is slightly misleading. The address gap represents the number of consecutive unused addresses available at all time.

Reply:
corrected -> https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/9522d0c0a5b0fd39ffab1b03235f364c78a972f3